File: crawl-paper/crawl-paper.js
import fs from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';
import dotenv from 'dotenv';
import { createClient } from '@supabase/supabase-js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Load biến môi trường từ thư mục RAG gốc
dotenv.config({ path: path.join(__dirname, '../.env') });

const supabaseUrl = process.env.SUPABASE_URL;
const supabaseKey = process.env.SUPABASE_SERVICE_ROLE_KEY;
if (!supabaseUrl || !supabaseKey) {
  console.error("❌ Không tìm thấy SUPABASE_URL hoặc SUPABASE_SERVICE_ROLE_KEY trong file .env!");
  process.exit(1);
}
const supabase = createClient(supabaseUrl, supabaseKey);
const OUTPUT_DIR = path.join(__dirname, 'document');
const METADATA_DIR = path.join(OUTPUT_DIR, 'metadata');

// Cấu hình tìm kiếm
const TARGET_TOTAL = 100; // Đã giảm xuống 100 để tránh bị ban IP khi tải PDF
const QUERIES = [
  { name: 'TurboQuant', query: 'all:"TurboQuant" OR all:"PolarQuant" OR all:"Quantized Johnson-Lindenstrauss"' },
  { name: 'ARQ', query: 'all:"Asynchronous RAG" OR all:"ARQ"' },
  { name: 'PQ', query: 'all:"Product Quantization" OR all:"Vector Quantization"' },
  { name: 'RAG', query: 'all:"Retrieval-Augmented Generation" OR all:"RAG"' }
];

async function delay(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

function cleanFilename(str) {
  return str.replace(/[\/\\?%*:|"<>]/g, '-').trim();
}

async function extractID(url) {
  const parts = url.split('/');
  return parts[parts.length - 1] || 'unknown';
}

async function crawlArxivWithPDF() {
  console.log(`🔍 Bắt đầu cào ~${TARGET_TOTAL} bài báo từ arXiv (KÈM THEO FILE PDF)...\n`);
  console.log(`⚠️ LƯU Ý: Quá trình sẽ chậm hơn đáng kể do cần tải trọn bộ file PDF (vài MB/file).`);
  console.log(`⚠️ arXiv có chính sách chống spam, script sẽ delay 2s mỗi bài báo để tránh bị ban IP.\n`);
  
  await fs.mkdir(METADATA_DIR, { recursive: true });
  console.log(`📁 Thư mục lưu trữ metadata: ${METADATA_DIR} \n   (PDF sẽ được tải thẳng lên hệ thống Supabase Storage)\n`);

  let totalSaved = 0;
  const savedIds = new Set();
  const maxPerQuery = 50; // Tổng 4 chủ đề x 50 bài = 200 bài

  for (const q of QUERIES) {
    if (totalSaved >= TARGET_TOTAL) break;

    console.log(`\n⏳ Đang tìm kiếm chủ đề: [${q.name}]...`);
    const encoded = encodeURIComponent(q.query);
    const apiUrl = `http://export.arxiv.org/api/query?search_query=${encoded}&start=0&max_results=${maxPerQuery}&sortBy=relevance&sortOrder=descending`;
    
    try {
      const res = await fetch(apiUrl);
      const xml = await res.text();
      const entries = xml.split('<entry>').slice(1);
      
      let countForQuery = 0;

      for (const entry of entries) {
        if (totalSaved >= TARGET_TOTAL) break;

        const titleMatch = entry.match(/<title>([\s\S]*?)<\/title>/);
        const summaryMatch = entry.match(/<summary>([\s\S]*?)<\/summary>/);
        const linkMatch = entry.match(/<id>([\s\S]*?)<\/id>/);
        const publishedMatch = entry.match(/<published>([\s\S]*?)<\/published>/);
        const pdfUrlMatch = entry.match(/<link title="pdf" href="([^"]+)"/);

        // Trích xuất tác giả
        const authors = [];
        const authorRegex = /<author>\s*<name>([\s\S]*?)<\/name>\s*<\/author>/g;
        let match;
        while ((match = authorRegex.exec(entry)) !== null) {
          authors.push(match[1].trim());
        }

        const link = linkMatch ? linkMatch[1].trim() : 'unknown';
        const id = await extractID(link);
        const pdfUrl = pdfUrlMatch ? pdfUrlMatch[1].replace('http://', 'https://') + '.pdf' : `https://arxiv.org/pdf/${id}.pdf`;
        
        if (savedIds.has(id)) continue;
        savedIds.add(id);

        const title = titleMatch ? titleMatch[1].replace(/\n/g, ' ').replace(/\s+/g, ' ').trim() : 'Untitled';
        const summary = summaryMatch ? summaryMatch[1].replace(/\n/g, ' ').replace(/\s+/g, ' ').trim() : '';
        const published = publishedMatch ? publishedMatch[1].split('T')[0] : 'Unknown';
        const baseFilename = `${id}_${cleanFilename(title).substring(0, 50)}`;

        const pdfFilename = `${baseFilename}.pdf`;
        
        // Tạo biến Storage URL công khai (Mặc định trước theo Supabase)
        const { data: publicUrlData } = supabase.storage.from('papers').getPublicUrl(pdfFilename);
        const supabaseUrlLink = publicUrlData.publicUrl;

        const txtPath = path.join(METADATA_DIR, `${baseFilename}.txt`);

        // ==========================================
        // TẢI RAW PDF & UPLOAD CLOUD
        // ==========================================
        console.log(` 📥 Đang tải PDF: ${id} (${countForQuery + 1}) - ${title.substring(0, 30)}...`);
        try {
          const pdfRes = await fetch(pdfUrl);
          if (pdfRes.ok) {
            const pdfBuffer = Buffer.from(await pdfRes.arrayBuffer());
            // Upload Storage
            console.log(`   ⬆️ Đang Upload PDF gốc lên Supabase Storage (Bucket: papers)...`);
            const { error: uploadError } = await supabase.storage
              .from('papers')
              .upload(pdfFilename, pdfBuffer, {
                contentType: 'application/pdf',
                upsert: true
              });
              
            if (uploadError) {
              console.log(`   ❌ Lỗi Upload Supabase: ${uploadError.message}`);
            } else {
              console.log(`   ✅ Đã chép file lên Supabase thành công.`);
              
              // Chỉ lưu metadata file khi PDF đã upload thành công
              const fileContent = 
`Tiêu đề bài báo: ${title}
Tác giả: ${authors.join(', ')}
Ngày xuất bản: ${published}
Nền tảng: arXiv
Link gốc: ${link}
Link PDF: ${pdfUrl}
Link Lưu Trữ (Supabase): ${supabaseUrlLink}
Chủ đề Crawler: ${q.name}

--- TÓM TẮT (ABSTRACT) ---
${summary}
`;
              await fs.writeFile(txtPath, fileContent, 'utf8');
              console.log(`   ✅ Đã tạo file metadata thành công.`);
            }
          } else {
            console.log(`   ❌ Bỏ qua PDF này (lỗi HTTP ${pdfRes.status})`);
          }
        } catch (downloadErr) {
          console.log(`   ❌ Bỏ qua PDF này (lỗi mạng)`);
        }

        countForQuery++;
        totalSaved++;

        // Delay 2s để tránh bị block ban IP khi request file PDF liên tục (arXiv giới hạn rất gắt PDF crawler)
        await delay(2000);
      }

      console.log(`✅ Đã cào xong ${countForQuery} bài báo + PDF cho [${q.name}]. Tổng thư viện: ${totalSaved}/${TARGET_TOTAL}`);
    } catch (err) {
      console.error(`❌ Lỗi API khi tìm [${q.name}]:`, err.message);
    }
  }

  console.log(`\n🎉 HOÀN THÀNH TOÀN BỘ!`);
  console.log(`Đã tải về TẤT CẢ FILE .TXT và .PDF vào thư mục:`);
  console.log(`👉 ${OUTPUT_DIR}`);
}

crawlArxivWithPDF();

File: backend/crawl_paper.py
import os
import time
import asyncio
import httpx
import re
import urllib.parse
from pathlib import Path
from dotenv import load_dotenv
from supabase import create_client, Client
import xml.etree.ElementTree as ET
import logging

# Thiết lập Logging cho UI
logger = logging.getLogger("Crawler")
logging.basicConfig(level=logging.INFO)

# Load environment variables
load_dotenv(dotenv_path=Path(__file__).parent.parent / ".env")

SUPABASE_URL = os.getenv("SUPABASE_URL")
SUPABASE_KEY = os.getenv("SUPABASE_SERVICE_ROLE_KEY")

if not SUPABASE_URL or not SUPABASE_KEY:
    print("❌ Không tìm thấy SUPABASE_URL hoặc SUPABASE_SERVICE_ROLE_KEY!")
    exit(1)

supabase: Client = create_client(SUPABASE_URL, SUPABASE_KEY)
METADATA_DIR = Path(__file__).parent / "document" / "metadata"

TARGET_TOTAL = 1100
QUERIES = [
    {"name": 'TurboQuant', "query": 'all:"TurboQuant" OR all:"PolarQuant" OR all:"Quantized Johnson-Lindenstrauss"'},
    {"name": 'ARQ', "query": 'all:"Asynchronous RAG" OR all:"ARQ" OR all:"Adaptive Reranking"'},
    {"name": 'PQ', "query": 'all:"Product Quantization" OR all:"Vector Quantization" OR all:"Product quantization for nearest neighbor search"'},
    {"name": 'RAG', "query": 'all:"Retrieval-Augmented Generation" OR all:"RAG system" OR all:"Knowledge-intensive NLP"'},
    {"name": 'ML_Optimization', "query": 'all:"Machine Learning Optimization" OR all:"Model Compression" OR all:"Vector Database"'},
    {"name": 'LLM_Inference', "query": 'all:"Large Language Model Inference" OR all:"Efficient LLM" OR all:"KV Cache"'}
]

def clean_filename(s: str) -> str:
    return re.sub(r'[\/\\?%*:|"<>]', '-', s).strip()

def extract_id(url: str) -> str:
    # Trích xuất ID từ URL dạng http://arxiv.org/abs/2203.08381v1 -> 2203.08381v1
    if not url: return "unknown"
    return url.split('/')[-1]

async def crawl_arxiv():
    logger.info(f"🔍 Bắt đầu cào ~{TARGET_TOTAL} bài báo từ arXiv...")
    METADATA_DIR.mkdir(parents=True, exist_ok=True)
    
    # Đảm bảo bucket 'papers' tồn tại
    try:
        buckets = supabase.storage.list_buckets()
        if not any(b.name == 'papers' for b in buckets):
            logger.info("⚠️ Bucket 'papers' chưa tồn tại. Đang tạo...")
            supabase.storage.create_bucket('papers', options={'public': True})
    except Exception as e:
        logger.warning(f"⚠️ Cảnh báo khi kiểm tra bucket: {e}")

    total_saved = 0
    saved_ids_in_session = set()
    max_per_query = 300
    headers = {
        "User-Agent": "Mozilla/5.0 (compatible; ArXivCrawler/2.0; +https://github.com/your-repo)",
        "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8"
    }

    # 1. Lấy danh sách ID đã tồn tại trong DB để tránh tải trùng
    try:
        res_db = supabase.table("papers").select("id").execute()
        existing_ids_in_db = {row['id'] for row in res_db.data}
        logger.info(f"💾 Database hiện có {len(existing_ids_in_db)} bài. Sẽ tự động bỏ qua các bài này.")
    except Exception as e:
        existing_ids_in_db = set()
        logger.warning(f"⚠️ Không thể kiểm tra ID từ DB: {e}")

    async with httpx.AsyncClient(timeout=60.0, follow_redirects=True, headers=headers) as client:
        for q in QUERIES:
            if total_saved >= TARGET_TOTAL: break
            
            logger.info(f"\n⏳ Đang tìm kiếm chủ đề: [{q['name']}]...")
            count_for_topic = 0
            
            # 2. Phân trang: Mỗi lần lấy 100 kết quả, lặp lại cho đến max_per_query
            for start_index in range(0, max_per_query, 100):
                if total_saved >= TARGET_TOTAL: break
                
                logger.info(f"   📑 Trang {start_index // 100 + 1} (start={start_index})...")
                encoded_query = urllib.parse.quote(q['query'])
                api_url = f"https://export.arxiv.org/api/query?search_query={encoded_query}&start={start_index}&max_results=100&sortBy=relevance&sortOrder=descending"
                
                try:
                    res = await client.get(api_url)
                    if res.status_code == 429:
                        logger.warning("   ⚠️ Rate limit API (429). Nghỉ 60s...")
                        await asyncio.sleep(60)
                        res = await client.get(api_url)
                    
                    res.raise_for_status()
                    root = ET.fromstring(res.text)
                    entries = root.findall('{http://www.w3.org/2005/Atom}entry')
                    
                    if not entries:
                        logger.info("   ℹ️ Không còn kết quả cho chủ đề này.")
                        break

                    for entry in entries:
                        if total_saved >= TARGET_TOTAL: break
                        
                        ns = {'atom': 'http://www.w3.org/2005/Atom'}
                        
                        id_elem = entry.find('atom:id', ns)
                        link_id = id_elem.text.strip() if id_elem is not None else ""
                        arxiv_id = extract_id(link_id)
                        
                        # KIỂM TRA TRÙNG LẶP
                        if arxiv_id in existing_ids_in_db or arxiv_id in saved_ids_in_session:
                            continue

                        title_elem = entry.find('atom:title', ns)
                        title = title_elem.text.replace('\n', ' ').strip() if title_elem is not None else 'Untitled'
                        
                        summary_elem = entry.find('atom:summary', ns)
                        summary = summary_elem.text.replace('\n', ' ').strip() if summary_elem is not None else ''
                        
                        published_elem = entry.find('atom:published', ns)
                        published = published_elem.text.split('T')[0] if published_elem is not None else 'Unknown'
                        
                        authors = [a.find('atom:name', ns).text.strip() for a in entry.findall('atom:author', ns) if a.find('atom:name', ns) is not None]

                        # 3. Tải PDF (Dùng link trực tiếp, ArXiv thường redirect .pdf về không đuôi)
                        pdf_url = f"https://export.arxiv.org/pdf/{arxiv_id}"
                        safe_title = clean_filename(title)[:50]
                        pdf_filename = f"{arxiv_id}_{safe_title}.pdf"
                        supabase_url_link = f"{SUPABASE_URL}/storage/v1/object/public/papers/{pdf_filename}"

                        logger.info(f"      📥 Đang tải: {arxiv_id}...")
                        
                        try:
                            # Nghỉ ngẫu nhiên từ 8-12 giây (để đạt tiến độ ~1100 bài / 3 tiếng)
                            import random
                            delay = random.uniform(8, 12)
                            await asyncio.sleep(delay)
                            
                            pdf_res = await client.get(pdf_url, timeout=45.0)
                            
                            # Xử lý Rate Limit khi tải PDF
                            if pdf_res.status_code == 429:
                                logger.warning("      ⚠️ Bị chặn (429) khi tải PDF. Nghỉ 120s để 'xả'...")
                                await asyncio.sleep(120)
                                pdf_res = await client.get(pdf_url, timeout=45.0)

                            if pdf_res.status_code == 200:
                                # 4. Upload lên Supabase Storage
                                upload_res = supabase.storage.from_("papers").upload(
                                    path=pdf_filename,
                                    file=pdf_res.content,
                                    file_options={"content-type": "application/pdf", "upsert": "true"}
                                )
                                
                                if hasattr(upload_res, 'error') and upload_res.error:
                                    logger.error(f"      ❌ Lỗi Storage: {upload_res.error}")
                                    continue

                                # 5. Lưu Metadata vào Database
                                try:
                                    supabase.table("papers").upsert({
                                        "id": arxiv_id,
                                        "title": title,
                                        "topic": q['name'],
                                        "url": pdf_url,
                                        "is_embedded": False
                                    }).execute()
                                    
                                    # Lưu Metadata TXT cục bộ làm backup
                                    txt_content = f"Tiêu đề: {title}\nTác giả: {', '.join(authors)}\nNgày: {published}\nLink: {pdf_url}\nChủ đề: {q['name']}\n\nAbstract: {summary}"
                                    with open(METADATA_DIR / f"{arxiv_id}.txt", "w", encoding="utf-8") as f:
                                        f.write(txt_content)

                                    saved_ids_in_session.add(arxiv_id)
                                    count_for_topic += 1
                                    total_saved += 1
                                    logger.info(f"      ✅ [OK] Đã lưu {arxiv_id}. Tổng mới: {total_saved}")
                                    
                                except Exception as db_err:
                                    logger.error(f"      ❌ Lỗi Database: {db_err}")
                            else:
                                logger.error(f"      ❌ PDF Error {pdf_res.status_code} cho {arxiv_id}")
                                
                        except Exception as e:
                            logger.error(f"      ❌ Lỗi kết nối PDF {arxiv_id}: {e}")

                except Exception as e:
                    logger.error(f"❌ Lỗi API ArXiv tại start={start_index}: {e}")
                    break
            
            logger.info(f"🏁 Đã xong chủ đề [{q['name']}]. Cào được {count_for_topic} bài mới.")

    logger.info(f"\n🎉 HOÀN THÀNH! Tổng cộng đã thêm mới {total_saved} bài báo.")

if __name__ == "__main__":
    asyncio.run(crawl_arxiv())
